### PR TITLE
Reorderable drag and drop component

### LINKED ISSUE #764: Issue when reordering layers
## Feature Description

Was able to move a layer to a new position, but unable to move it back to original position
Vid - https://drive.google.com/file/d/16MtsHg3dfmrGAFhvM29zYUqoBJKoXqQW/view
Taken from Claudio in #584 

I did some additional testing and there seem to be issues with reordering layers. 
I am never able to drag a layer to the very top layer making it the new top. 
Other times I just can drag a layer to another spot, but I can't see a pattern as I've tried multiple variations of positions and element types. 

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by wassgha**
Closes #764
Partial to #746
Partial to #332
Partial to #517

This refactors/rethinks the layer re-ordering component into three generic components: `<Reorderable>` (a container of sortable elements), `<ReorderableSeparator>` and `<ReorderableItem>`. The refactor is currently only used in the layers panel, in a second PR it will be used for the `GridView` and the carousel thumbnails in order to deprecate the current `DropZone` component.

/cc @dvoytenko @barklund

### CODE REVIEW

File: assets/src/edit-story/components/panels/layer/panel.js [hunk 1]
```diff
@@ -23,13 +23,19 @@ import { __ } from '@wordpress/i18n';
...
+import useLayers from './useLayers';
 
 function LayerPanel() {
+  const layers = useLayers();
+
   return (
-    <Panel name="layers" initialHeight={DEFAULT_LAYERS_VISIBLE * LAYER_HEIGHT}>
+    <Panel
+      name="layers"
+      initialHeight={(layers?.length || DEFAULT_LAYERS_VISIBLE) * LAYER_HEIGHT}
```
Comment (swissspidy) [inline assets/src/edit-story/components/panels/layer/panel.js]: These are all changes from #704?

Comment (wassgha) [reply assets/src/edit-story/components/panels/layer/panel.js]: Yes! Please only review after that one

**Review by dvoytenko — APPROVED** [review]
The only question is whether we should get https://github.com/google/web-stories-wp/pull/551 in for this.

**Comment by wassgha** [comment]
Up to @barklund

File: assets/src/edit-story/components/panels/layer/panel.js [hunk 2]
```diff
@@ -23,13 +23,19 @@ import { __ } from '@wordpress/i18n';
  * Internal dependencies
  */
 import { Panel, PanelTitle, PanelContent } from '../panel';
-import { DEFAULT_LAYERS_VISIBLE, LAYER_HEIGHT } from './constants';
+import { LAYER_HEIGHT, DEFAULT_LAYERS_VISIBLE } from './constants';
```
Comment (barklund) [inline assets/src/edit-story/components/panels/layer/panel.js:26]: Much better order 😉

File: assets/src/edit-story/components/panels/layer/panel.js (hunk 1 above)
Comment (barklund) [inline assets/src/edit-story/components/panels/layer/panel.js]: Is this aligned with @samitron7? The design originally called for always 6 initial rows. We have #227 too, by the way, which is then fixed by this issue?

File: assets/src/edit-story/components/panels/panel/panel.js [hunk 3]
```diff
@@ -35,9 +35,34 @@ const Wrapper = styled.section`
 function Panel({ initialHeight, name, children }) {
   const [isCollapsed, setIsCollapsed] = useState(false);
   const [height, setHeight] = useState(initialHeight);
+  const [manuallyChanged, setManuallyChanged] = useState(false);
 
   const collapse = useCallback(() => setIsCollapsed(true), []);
-  const expand = useCallback(() => setIsCollapsed(false), []);
+  const expand = useCallback(() => {
+    setIsCollapsed(false);
+    setHeight(initialHeight);
```
Comment (barklund) [inline assets/src/edit-story/components/panels/panel/panel.js]: Controversial, but fair enough 😄

File: assets/src/edit-story/components/panels/panel/panel.js [hunk 4]
```diff
@@ -35,9 +35,34 @@ const Wrapper = styled.section`
...
+    }
+    setHeight(initialHeight);
+  }, [manuallyChanged, initialHeight]);
+
+  const manuallySetHeight = useCallback(
+    (h) => {
+      setManuallyChanged(true);
+      setHeight(h);
+    },
+    [setManuallyChanged, setHeight]
```
Comment (barklund) [inline assets/src/edit-story/components/panels/panel/panel.js]: These deps aren't necessary (nor required by linter).

File: assets/src/edit-story/components/reorderable/reorderable.js [hunk 5]
```diff
@@ -99,13 +110,50 @@ function useLayerReordering(layer) {
...
+  );
 }
 
-export default useLayerReordering;
+Reorderable.propTypes = {
+  children: PropTypes.oneOfType([
+    PropTypes.arrayOf(PropTypes.node),
+    PropTypes.node,
+  ]).isRequired,
+  onPositionChange: PropTypes.func,
```
Comment (barklund) [inline assets/src/edit-story/components/reorderable/reorderable.js]: Shouldn't we require this?

File: assets/src/edit-story/components/panels/layer/provider.js [hunk 6]
```diff
@@ -28,17 +28,14 @@ import useLayers from './useLayers';
 
 function LayerProvider({ children }) {
```
Comment (barklund) [inline assets/src/edit-story/components/panels/layer/provider.js]: Is this file still doing anything?

File: assets/src/edit-story/components/panels/layer/layerList.js [hunk 7]
```diff
@@ -18,60 +18,69 @@
...
 
-const REORDER_MESSAGE = __(
-  /* translators: d: new layer position. */
-  'Reordering layers. Press Escape to abort. Release mouse to drop in position %d.',
-  'web-stories'
-);
-
-const LayerList = styled.div.attrs({ role: 'listbox' })`
-  display: flex;
+const LayerList = styled(Reorderable)`
```
Comment (barklund) [inline assets/src/edit-story/components/panels/layer/layerList.js]: Might want to put `.attrs({'aria-orientation': 'vertical'})` on here too, though [it's default](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/listbox_role).

**Review by barklund — COMMENTED** [review]
Works great, just a few nits. Let's get this in first, I'll rework #551 in this new generalized context.

File: assets/src/edit-story/components/panels/layer/panel.js (hunk 1 above)
Comment (wassgha) [reply assets/src/edit-story/components/panels/layer/panel.js]: Tagging @samitron7 here but from our discussion it seemed like she wanted it to vary by number of items more than page size

File: assets/src/edit-story/components/panels/layer/provider.js (hunk 6 above)
Comment (wassgha) [reply assets/src/edit-story/components/panels/layer/provider.js]: Nope! Deleted

**Review by barklund — APPROVED** [review]
LGTM

**Pull request merged**